cpp/src/parquet/arrow/writer.cc

```diff
@@ -462,7 +462,8 @@ class FileWriterImpl : public FileWriter {
       offset += batch_size;
 
       // Flush current row group if it is full.
-      if (row_group_writer_->num_rows() >= max_row_group_length) {
+      if (row_group_writer_->num_rows() >= max_row_group_length &&
+          offset < batch.num_rows()) {
```

Comment:
Why not avoid `WriteBatch` above instead?

Reply:
Hmmm `NewBufferedRowGroup()` here is as "if we have more rows needs to write, we need to switch a buffered-writer", and this is better to put after the write? What would you expect `WriteBatch` and `NewBufferedRowGroup` to be like?

Reply:
Yeah, the main reason is because `NewBufferedRowGroup()` is "close if having a writer and create a new buffered writer", previously, if `offset == batch.num_rows()`, it will allocate an empty writer. When file is closed then, it will leave an empty row-group

`WriteTable` doesn't have this problem because it extract a lambda expression `WriteRowGroup`, it just "close previously writer, write the records", and waiting for next `WriteRowGroup` to close the previous writer